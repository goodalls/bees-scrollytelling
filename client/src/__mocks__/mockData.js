module.exports = {
  appDefaultState: {
    scrollPosition: 750,
    questionsEnabled: true,
    currentUserId: null,
    displayGraphCover: true
  },
  event: {
    target: {
      name: 'Jeff',
      value: 'websocket'
    }
  },
  beeFactsPost: {
    body: '{"users_id":2,"user_answer":"","question":"bee species"}',
    headers: { 'Content-type': 'application/json' },
    method: 'POST'
  },
  headerDefaultState: {
    concern: '',
    age: '',
    location: {},
    answered: false
  },
  impactDefaultState: {
    percent: 0,
    unit: 'million',
    monetary: '',
    economy: '',
    answeredeconomy: false,
    answeredpercent: false
  },
  impactFactsPost: {
    body: '{"users_id":2,"question":"are plants great"}',
    headers: { 'Content-type': 'application/json' },
    method: 'POST'
  },
  problemsPostBody: {
    method: 'POST',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify({
      users_id: 2,
      user_answer: 30,
      question: 'What percent of bee colonies do beekeepers lose every year?'
    })
  },
  rawStatistics: [
    {
      id: 48,
      question:
        'What percentage of your diet do you think is provided by bees?',
      user_answer: '35',
      users_id: 50
    },
    {
      id: 49,
      question: 'How much do bees contribute to the economy, anually?',
      user_answer: '400_million',
      users_id: 50
    },
    { id: 50, question: 'bee species', user_answer: '200', users_id: 50 },
    {
      id: 51,
      question: 'What percent of bee colonies do beekeepers lose every year?',
      user_answer: '40',
      users_id: 50
    },
    {
      id: 52,
      question:
        'What percentage of your diet do you think is provided by bees?',
      user_answer: '80',
      users_id: 51
    },
    {
      id: 53,
      question: 'How much do bees contribute to the economy, anually?',
      user_answer: '400_billion',
      users_id: 51
    },
    { id: 54, question: 'bee species', user_answer: '100', users_id: 51 },
    {
      id: 55,
      question: 'What percent of bee colonies do beekeepers lose every year?',
      user_answer: '25',
      users_id: 51
    }
  ],
  cleanedStatistics: {
    'How much do bees contribute to the economy, anually?': {
      count: { '400_million': 1, '400_billion': 1 },
      sum: 400400000000
    },
    'What percent of bee colonies do beekeepers lose every year?': {
      count: { 25: 1, 40: 1 },
      sum: 65
    },
    'What percentage of your diet do you think is provided by bees?': {
      count: { 35: 1, 80: 1 },
      sum: 115
    },
    'bee species': { count: { 100: 1, 200: 1 }, sum: 300 }
  },
  cleanedAverageData: {
    'How much do bees contribute to the economy, anually?': {
      count: { '400_million': 1, '400_billion': 1 },
      sum: 400400000000,
      average: 200200000000
    },
    'What percent of bee colonies do beekeepers lose every year?': {
      count: { 25: 1, 40: 1 },
      sum: 65,
      average: 32.5
    },
    'What percentage of your diet do you think is provided by bees?': {
      count: { 35: 1, 80: 1 },
      sum: 115,
      average: 57.5
    },
    'bee species': { count: { 100: 1, 200: 1 }, sum: 300, average: 150 }
  },
  pieChartDefaultState: {
    class: '',
    degree: 0,
    percent: 10
  },
  percentEvent: {
    target: {
      name: 'percent',
      value: 55
    }
  }
};
