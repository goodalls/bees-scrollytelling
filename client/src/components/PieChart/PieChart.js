import React, { Component } from 'react';
import '../../assets/variables.scss';
import './PieChart.css';

class PieChart extends Component {
  constructor() {
    super();
    this.state = {
      class: '',
      degree: 0,
      percent: 10
    }
  }

  componentDidMount = () => {
    this.updateChart();
  }

  updateChart = () => {
    const degree = 360*this.state.percent/100;
    if(this.state.percent > 50) {
      this.setState({ class: 'gt-50', degree })
    } else {
      this.setState({ class: '', degree })
    }
  }

  changePercent = async (e) => {
    await this.setState({ [e.target.name]: e.target.value })
    this.updateChart();
    this.props.updateImpactPercent(this.state.percent);
  }

  render() {
    return (
      <div className='pie-chart-cont'>
        <div className={`progress-pie-chart ${ this.state.class }`} 
             data-percent={ this.state.percent }>
          <div className='ppc-progress'>
            <div className='ppc-progress-fill' 
                 style={{transform: `rotate(${this.state.degree}deg)`}}></div>
          </div>
          <div className='ppc-percents'>
            <div className='pcc-percents-wrapper'>
              <span>{ this.state.percent }%</span>
            </div>
          </div>
        </div>

        <div>
          <label>Enter a Percentage</label>
          <input 
            name='percent' 
            type='number' 
            placeholder='Percent' 
            step='5' 
            min='0' 
            max='100'
            value={ this.state.percent }
            onChange={ this.changePercent }
          />
        </div>
      </div>
    )
  }
}

export default PieChart;