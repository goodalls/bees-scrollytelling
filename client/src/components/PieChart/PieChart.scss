@import '../../assets/variables.scss';

@mixin circle($size) {
  content: "";
  position: absolute;
  border-radius: 50%;
  left: calc(50% - #{$size/2});
  top: calc(50% - #{$size/2});
  width: $size;
  height: $size;
}

$size: 200px;
.progress-pie-chart {
  width: $size;
  height: $size;
  border-radius: 50%;
  background-color: #E5E5E5;
  position: relative;
  &.gt-50 {
     background-color: $seafoam;
  }
}
.ppc-progress {
  @include circle($size);
  clip: rect(0, $size, $size, #{$size/2});
  .ppc-progress-fill {
    @include circle($size);
    clip: rect(0, #{$size/2}, $size, 0);
    background: $seafoam;
    transform: rotate(60deg);
  }
  .gt-50 & {
    clip: rect(0, #{$size/2}, $size, 0);
    .ppc-progress-fill {
      clip: rect(0, $size, $size, #{$size/2});
      background: #E5E5E5;
    }
  }
}
.ppc-percents {
  @include circle(#{$size/1.15});
  background: #fff;
  text-align: center;
  display: table;
  span {
    display: block;
    font-size: 2.6em;
    font-weight: bold;
    color: $seafoam;
  }
}
.pcc-percents-wrapper {
  display: table-cell;
  vertical-align: middle;
}

.pie-chart-cont {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 45px 0;
}

.pie-chart-cont label {
  padding-left: 16px;
}

.pie-chart-cont input {
  margin-left: 20px;
  border: 2px solid #e5e5e5;
  width: 90px;
  border-radius: 52px;
  height: 37px;
  padding-left: 5px;
  font-size: 1.3em;
  text-align: center;
  z-index: 3;
}

.progress-pie-chart {
  margin: 10px 0;
}