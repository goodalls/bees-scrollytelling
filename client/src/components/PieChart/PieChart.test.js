import React from 'react';
import PieChart from './PieChart';
import { shallow } from 'enzyme';
import mockData from '../../__mocks__/mockData';

describe("PieChart", () => {
  let wrapper;
  let updateImpactPercent = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<PieChart updateImpactPercent={updateImpactPercent} />);
  });

  it('matches the Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has default state', () => {
    wrapper = shallow(<PieChart updateImpactPercent={updateImpactPercent} />, { disableLifecycleMethods: true });

    expect(wrapper.instance().state).toEqual(mockData.pieChartDefaultState);
  });


  describe('updateChart', async () => {

    it('updates state with class gt-50 if the percent is above 50', async() => {
      await wrapper.instance().setState({ percent: 80 });
      await wrapper.instance().updateChart();

      expect(wrapper.instance().state.class).toEqual('gt-50');
      expect(wrapper.instance().state.degree).toEqual(288);
    });

    it('updates state with class if the percent is less than 50', async() => {
      await wrapper.instance().updateChart();

      expect(wrapper.instance().state.class).toEqual('');
      expect(wrapper.instance().state.degree).toEqual(36);
    });
  });

  describe('changePercent', () => {
    it('updates state with the event name and target', async() => {
      wrapper.instance().updateChart = jest.fn();
      await wrapper.instance().changePercent(mockData.percentEvent);
      
      expect(wrapper.instance().state.percent).toEqual(55);
    });

    it('calls updateChart', async() => {
      wrapper.instance().updateChart = jest.fn();
      await wrapper.instance().changePercent(mockData.percentEvent);

      expect(wrapper.instance().updateChart).toHaveBeenCalled();
    });

    it('calls updateImpactPercent', async() => {
      await wrapper.instance().changePercent(mockData.percentEvent);

      expect(updateImpactPercent).toHaveBeenCalledWith(55);
    })
  });
})