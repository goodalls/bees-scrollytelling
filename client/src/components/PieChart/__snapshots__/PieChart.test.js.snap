// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart matches the Snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PieChart
    updateImpactPercent={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="progress-pie-chart "
          data-percent={10}
>
          <div
                    className="ppc-progress"
          >
                    <div
                              className="ppc-progress-fill"
                              style={
                                        Object {
                                                  "transform": "rotate(36deg)",
                                                }
                              }
                    />
          </div>
          <div
                    className="ppc-percents"
          >
                    <div
                              className="pcc-percents-wrapper"
                    >
                              <span>
                                        10
                                        %
                              </span>
                    </div>
          </div>
</div>,
        <div>
          <label>
                    Enter a Percentage
          </label>
          <input
                    max="100"
                    min="0"
                    name="percent"
                    onChange={[Function]}
                    placeholder="Percent"
                    step="5"
                    type="number"
                    value={10}
          />
</div>,
      ],
      "className": "pie-chart-cont",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="ppc-progress"
>
              <div
                            className="ppc-progress-fill"
                            style={
                                          Object {
                                                        "transform": "rotate(36deg)",
                                                      }
                            }
              />
</div>,
            <div
              className="ppc-percents"
>
              <div
                            className="pcc-percents-wrapper"
              >
                            <span>
                                          10
                                          %
                            </span>
              </div>
</div>,
          ],
          "className": "progress-pie-chart ",
          "data-percent": 10,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="ppc-progress-fill"
                style={
                                Object {
                                                "transform": "rotate(36deg)",
                                              }
                }
/>,
              "className": "ppc-progress",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "ppc-progress-fill",
                "style": Object {
                  "transform": "rotate(36deg)",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="pcc-percents-wrapper"
>
                <span>
                                10
                                %
                </span>
</div>,
              "className": "ppc-percents",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  10
                  %
</span>,
                "className": "pcc-percents-wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    10,
                    "%",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  10,
                  "%",
                ],
                "type": "span",
              },
              "type": "div",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Enter a Percentage
</label>,
            <input
              max="100"
              min="0"
              name="percent"
              onChange={[Function]}
              placeholder="Percent"
              step="5"
              type="number"
              value={10}
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Enter a Percentage",
            },
            "ref": null,
            "rendered": "Enter a Percentage",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "max": "100",
              "min": "0",
              "name": "percent",
              "onChange": [Function],
              "placeholder": "Percent",
              "step": "5",
              "type": "number",
              "value": 10,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="progress-pie-chart "
            data-percent={10}
>
            <div
                        className="ppc-progress"
            >
                        <div
                                    className="ppc-progress-fill"
                                    style={
                                                Object {
                                                            "transform": "rotate(36deg)",
                                                          }
                                    }
                        />
            </div>
            <div
                        className="ppc-percents"
            >
                        <div
                                    className="pcc-percents-wrapper"
                        >
                                    <span>
                                                10
                                                %
                                    </span>
                        </div>
            </div>
</div>,
          <div>
            <label>
                        Enter a Percentage
            </label>
            <input
                        max="100"
                        min="0"
                        name="percent"
                        onChange={[Function]}
                        placeholder="Percent"
                        step="5"
                        type="number"
                        value={10}
            />
</div>,
        ],
        "className": "pie-chart-cont",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="ppc-progress"
>
                <div
                                className="ppc-progress-fill"
                                style={
                                                Object {
                                                                "transform": "rotate(36deg)",
                                                              }
                                }
                />
</div>,
              <div
                className="ppc-percents"
>
                <div
                                className="pcc-percents-wrapper"
                >
                                <span>
                                                10
                                                %
                                </span>
                </div>
</div>,
            ],
            "className": "progress-pie-chart ",
            "data-percent": 10,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="ppc-progress-fill"
                  style={
                                    Object {
                                                      "transform": "rotate(36deg)",
                                                    }
                  }
/>,
                "className": "ppc-progress",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "ppc-progress-fill",
                  "style": Object {
                    "transform": "rotate(36deg)",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="pcc-percents-wrapper"
>
                  <span>
                                    10
                                    %
                  </span>
</div>,
                "className": "ppc-percents",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <span>
                    10
                    %
</span>,
                  "className": "pcc-percents-wrapper",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      10,
                      "%",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    10,
                    "%",
                  ],
                  "type": "span",
                },
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                Enter a Percentage
</label>,
              <input
                max="100"
                min="0"
                name="percent"
                onChange={[Function]}
                placeholder="Percent"
                step="5"
                type="number"
                value={10}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Enter a Percentage",
              },
              "ref": null,
              "rendered": "Enter a Percentage",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "max": "100",
                "min": "0",
                "name": "percent",
                "onChange": [Function],
                "placeholder": "Percent",
                "step": "5",
                "type": "number",
                "value": 10,
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
